packages/protocol/contracts/stability/Broker.sol

```diff
@@ -0,0 +1,197 @@
+
+  /**
+   * @notice Sets initialized == true on implementation contracts.
+   * @param test Set to true to skip implementation initialization.
+   */
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @notice Allows the contract to be upgradable via the proxy.
+   * @param _exchangeManager The address of the PairManager contract.
+   * @param _reserve The address of the Rezerve contract.
+   */
+  function initilize(address _exchangeManager, address _reserve) external initializer {
+    _transferOwnership(msg.sender);
+    addExchangeManager(_exchangeManager);
+    setReserve(_reserve);
+  }
+
+  /* ==================== Mutative Functions ==================== */
+
+  /**
+   * @notice Add exchange manager
+   * @param _exchangeManager The address of the exchange manager to add
+   * @return index The index where it was inserted
+   */
+  function addExchangeManager(address _exchangeManager) public onlyOwner returns (uint256 index) {
+    require(
+      !existsInExchangeManagers(_exchangeManager),
+      "ExchangeManager already exists in the list"
+    );
+    require(_exchangeManager != address(0), "ExchangeManager address can't be 0");
+    exchangeManagers.push(_exchangeManager);
+    emit ExchangeManagerAdded(_exchangeManager);
+    return index = exchangeManagers.length - 1;
+  }
+
+  /**
+   * @notice Remove an exchange manager at an index
+   * @param _exchangeManager The address of the exchange manager to remove
+   * @param index The index in the exchange managers array
+   * @return bool returns true if successful
+   */
+  function removeExchangeManager(address _exchangeManager, uint256 index)
+    public
+    onlyOwner
+    returns (bool)
+  {
+    require(
+      index < exchangeManagers.length && exchangeManagers[index] == _exchangeManager,
+      "index into exchangeManagers list not mapped to token"
+    );
+    exchangeManagers[index] = exchangeManagers[exchangeManagers.length - 1];
+    exchangeManagers.pop();
+    emit ExchangeManagerRemoved(_exchangeManager);
+    return true;
+  }
+
+  /**
+   * @notice Set the Mento reserve address
+   * @param _reserve The Mento reserve address
+   */
+  function setReserve(address _reserve) public onlyOwner {
+    require(_reserve != address(0), "Reserve address must be set");
+    emit ReserveSet(_reserve, address(reserve));
+    reserve = IReserve(_reserve);
+  }
+
+  /**
+   * @notice Calculate amountIn of tokenIn for a given amountIn of tokenIn
+   * @param exchangeId The id of the exchange to use
+   * @param tokenIn The token to be sold
+   * @param tokenOut The token to be bought
+   * @param amountOut The amount of tokenOut to be bought
+   * @return amountIn The amount of tokenIn to be sold
+   */
+  function getAmountIn(bytes32 exchangeId, address tokenIn, address tokenOut, uint256 amountOut)
+    external
+    returns (uint256 amountIn)
+  {
+    return amountOut = exchangeManager.getAmountIn(exchangeId, tokenIn, tokenOut, amountOut);
```

Comment:
This function should also take in an `address exchangeManager` as the first parameter, check if it's a registered exchangeManager and then cast that to an `IExchangeManager(exchangeManager)` and call getAmountIn on it.